File: pokersolver/hand_evaluator.py
"""
Hand evaluation for poker.
"""

from collections import Counter
from enum import IntEnum
from typing import List, Tuple
from pokersolver.card import Card


class HandRank(IntEnum):
    """Poker hand rankings."""
    HIGH_CARD = 1
    PAIR = 2
    TWO_PAIR = 3
    THREE_OF_A_KIND = 4
    STRAIGHT = 5
    FLUSH = 6
    FULL_HOUSE = 7
    FOUR_OF_A_KIND = 8
    STRAIGHT_FLUSH = 9


class HandEvaluator:
    """Evaluates poker hands."""
    
    @staticmethod
    def evaluate(cards: List[Card]) -> Tuple[int, List[int]]:
        """
        Evaluate a poker hand.
        
        Args:
            cards: List of cards (typically 5-7 cards for Texas Hold'em)
            
        Returns:
            Tuple of (hand_rank, kickers) where kickers are used for tie-breaking
        """
        if len(cards) < 5:
            raise ValueError("Need at least 5 cards to evaluate a hand")
        
        # For 7 cards, find the best 5-card combination
        if len(cards) == 7:
            return HandEvaluator._best_five_card_hand(cards)
        
        # Evaluate 5-card hand
        ranks = sorted([c.rank for c in cards], reverse=True)
        suits = [c.suit for c in cards]
        
        is_flush = len(set(suits)) == 1
        is_straight = HandEvaluator._is_straight(ranks)
        
        if is_straight and is_flush:
            return (HandRank.STRAIGHT_FLUSH, [max(ranks)])
        
        rank_counts = Counter(ranks)
        counts = sorted(rank_counts.items(), key=lambda x: (x[1], x[0]), reverse=True)
        
        # Four of a kind
        if counts[0][1] == 4:
            return (HandRank.FOUR_OF_A_KIND, [counts[0][0], counts[1][0]])
        
        # Full house
        if counts[0][1] == 3 and counts[1][1] == 2:
            return (HandRank.FULL_HOUSE, [counts[0][0], counts[1][0]])
        
        # Flush
        if is_flush:
            return (HandRank.FLUSH, ranks)
        
        # Straight
        if is_straight:
            # Handle A-2-3-4-5 straight (wheel)
            if ranks == [14, 5, 4, 3, 2]:
                return (HandRank.STRAIGHT, [5])
            return (HandRank.STRAIGHT, [max(ranks)])
        
        # Three of a kind
        if counts[0][1] == 3:
            kickers = [counts[1][0], counts[2][0]]
            return (HandRank.THREE_OF_A_KIND, [counts[0][0]] + kickers)
        
        # Two pair
        if counts[0][1] == 2 and counts[1][1] == 2:
            pairs = sorted([counts[0][0], counts[1][0]], reverse=True)
            return (HandRank.TWO_PAIR, pairs + [counts[2][0]])
        
        # One pair
        if counts[0][1] == 2:
            kickers = [c[0] for c in counts[1:]]
            return (HandRank.PAIR, [counts[0][0]] + kickers)
        
        # High card
        return (HandRank.HIGH_CARD, ranks)
    
    @staticmethod
    def _is_straight(ranks: List[int]) -> bool:
        """Check if ranks form a straight."""
        if len(set(ranks)) != 5:
            return False
        
        # Check for A-2-3-4-5 straight (wheel)
        if sorted(ranks) == [2, 3, 4, 5, 14]:
            return True
        
        # Check for regular straight
        return max(ranks) - min(ranks) == 4
    
    @staticmethod
    def _best_five_card_hand(cards: List[Card]) -> Tuple[int, List[int]]:
        """Find the best 5-card hand from 7 cards."""
        from itertools import combinations
        
        best_hand = None
        best_eval = (0, [])
        
        for combo in combinations(cards, 5):
            evaluation = HandEvaluator.evaluate(list(combo))
            if evaluation > best_eval:
                best_eval = evaluation
                best_hand = combo
        
        return best_eval
    
    @staticmethod
    def compare_hands(hand1: List[Card], hand2: List[Card]) -> int:
        """
        Compare two poker hands.
        
        Args:
            hand1: First hand
            hand2: Second hand
            
        Returns:
            1 if hand1 wins, -1 if hand2 wins, 0 if tie
        """
        eval1 = HandEvaluator.evaluate(hand1)
        eval2 = HandEvaluator.evaluate(hand2)
        
        if eval1 > eval2:
            return 1
        elif eval1 < eval2:
            return -1
        else:
            return 0
    
    @staticmethod
    def hand_rank_name(rank: int) -> str:
        """Get the name of a hand rank."""
        names = {
            HandRank.HIGH_CARD: "High Card",
            HandRank.PAIR: "Pair",
            HandRank.TWO_PAIR: "Two Pair",
            HandRank.THREE_OF_A_KIND: "Three of a Kind",
            HandRank.STRAIGHT: "Straight",
            HandRank.FLUSH: "Flush",
            HandRank.FULL_HOUSE: "Full House",
            HandRank.FOUR_OF_A_KIND: "Four of a Kind",
            HandRank.STRAIGHT_FLUSH: "Straight Flush",
        }
        return names.get(rank, "Unknown")

File: app/src/main/java/com/example/pokersolverGTO/equity/logic/HandEvaluator.kt
package com.example.pokersolverGTO.equity.logic

import com.example.pokersolverGTO.equity.models.*

class HandEvaluator {

    fun evaluateHand(cards: List<Card>): HandResult {
        require(cards.size >= 5) { "Need at least 5 cards to evaluate" }

        return if (cards.size == 7) {
            findBest5CardHand(cards)
        } else {
            evaluateFiveCards(cards.take(5))
        }
    }

    private fun findBest5CardHand(sevenCards: List<Card>): HandResult {
        val combinations = getCombinations(sevenCards, 5)
        return combinations.maxOf { evaluateFiveCards(it) }
    }

    private fun getCombinations(cards: List<Card>, k: Int): List<List<Card>> {
        if (k == 0) return listOf(emptyList())
        if (cards.isEmpty()) return emptyList()

        val head = cards.first()
        val tail = cards.drop(1)

        val withHead = getCombinations(tail, k - 1).map { listOf(head) + it }
        val withoutHead = getCombinations(tail, k)

        return withHead + withoutHead
    }

    private fun evaluateFiveCards(cards: List<Card>): HandResult {
        val sortedCards = cards.sortedByDescending { it.rank.value }
        val ranks = sortedCards.map { it.rank.value }
        val suits = sortedCards.map { it.suit }

        val isFlush = suits.toSet().size == 1
        val isStraight = checkStraight(ranks)

        val rankCounts = ranks.groupingBy { it }.eachCount()
        val counts = rankCounts.values.sortedDescending()

        return when {
            isStraight && isFlush && ranks.first() == 12 ->
                HandResult(HandRank.ROYAL_FLUSH, sortedCards, ranks)

            isStraight && isFlush ->
                HandResult(HandRank.STRAIGHT_FLUSH, sortedCards, ranks)

            counts == listOf(4, 1) -> {
                val quadRank = rankCounts.entries.first { it.value == 4 }.key
                val kicker = rankCounts.entries.first { it.value == 1 }.key
                HandResult(HandRank.FOUR_OF_A_KIND, sortedCards, listOf(quadRank, kicker))
            }

            counts == listOf(3, 2) -> {
                val tripRank = rankCounts.entries.first { it.value == 3 }.key
                val pairRank = rankCounts.entries.first { it.value == 2 }.key
                HandResult(HandRank.FULL_HOUSE, sortedCards, listOf(tripRank, pairRank))
            }

            isFlush ->
                HandResult(HandRank.FLUSH, sortedCards, ranks)

            isStraight ->
                HandResult(HandRank.STRAIGHT, sortedCards, ranks)

            counts == listOf(3, 1, 1) -> {
                val tripRank = rankCounts.entries.first { it.value == 3 }.key
                val kickers = rankCounts.entries.filter { it.value == 1 }.map { it.key }.sortedDescending()
                HandResult(HandRank.THREE_OF_A_KIND, sortedCards, listOf(tripRank) + kickers)
            }

            counts == listOf(2, 2, 1) -> {
                val pairs = rankCounts.entries.filter { it.value == 2 }.map { it.key }.sortedDescending()
                val kicker = rankCounts.entries.first { it.value == 1 }.key
                HandResult(HandRank.TWO_PAIR, sortedCards, pairs + kicker)
            }

            counts == listOf(2, 1, 1, 1) -> {
                val pairRank = rankCounts.entries.first { it.value == 2 }.key
                val kickers = rankCounts.entries.filter { it.value == 1 }.map { it.key }.sortedDescending()
                HandResult(HandRank.PAIR, sortedCards, listOf(pairRank) + kickers)
            }

            else ->
                HandResult(HandRank.HIGH_CARD, sortedCards, ranks)
        }
    }

    private fun checkStraight(ranks: List<Int>): Boolean {
        val sortedRanks = ranks.sorted()

        // Check for A-2-3-4-5 (wheel)
        if (sortedRanks == listOf(0, 1, 2, 3, 12)) return true

        // Check for consecutive ranks
        for (i in 0 until sortedRanks.size - 1) {
            if (sortedRanks[i + 1] != sortedRanks[i] + 1) return false
        }

        return true
    }
}

